File: Kotlin/src/Sphere.kt
import java.lang.Math.sqrt

data class Sphere(val center: Vector,
                  val radius2:Double,
                  val surface: Surface) : Thing {
    override fun normal(position: Vector) = position.minus(this.center).normalize()
    override fun intersect(ray: Ray): Intersection? {
        val eo = this.center - ray.start
        val v = eo.dot(ray.direction)

        var dist = 0.0
        if (v >= 0) {
            val disc = this.radius2 - (eo.dot(eo) - v * v)
            if (disc >= 0) {
                dist = v - sqrt(disc)
            }
        }
        return if (dist == 0.0) {
            null
        } else {
            Intersection( this, ray, dist )
        }
    }

    override fun surface() = this.surface
}

File: Go/src/ray4_goroutine/sphere.go
package main

import "math"
import . "github.com/go-gl/mathgl/mgl64"

type Sphere struct {
	Center  Vec3
	Radius2 float64
	surface Surface
}

func (s Sphere) normal(pos Vec3) (r Vec3) {
	return pos.Sub(s.Center).Normalize()
}

func (s Sphere) intersect(r Ray) (Intersection, bool) {
	eo := s.Center.Sub(r.start)
	v := eo.Dot(r.direction)
	var dist float64
	if v >= 0 {
		disc := s.Radius2 - (eo.Dot(eo) - v*v)
		if disc >= 0 {
			dist = v - float64(math.Sqrt(float64(disc)))
		}
	}
	if dist == 0 {
		return Intersection{}, false
	} else {
		return Intersection{thing: s, ray: r, distance: dist}, true
	}

}

func (s Sphere) Surface() Surface {
	return s.surface
}
